### PR TITLE
Add Entropy/Mode methods to discrete distributions

### PR DESCRIPTION
**Pull request by stergiosba**
Solves #1696 
Adds the following:

- `entropy` method for: Bernoulli/Categorical/DiscreteUniform/Geometric
- `mode` property for: Bernoulli/Categorical/Binomial/Poisson/Geometric
- `name` property for all distributions and explanation for NotImplementedError.

Current Status: Passes all local tests CPU and GPU (tested on JAX 0.4.23 and CUDA 12.2)
Added dependencies: None

### CODE REVIEW

**Comment by stergiosba** [comment]
@fehiepsi Hello!
Whenever you want and you are ready take a look.
Thank you!

File: numpyro/distributions/discrete.py [hunk 1]
```diff
@@ -385,6 +474,14 @@ def variance(self):
     def support(self):
         return constraints.integer_interval(0, jnp.shape(self.logits)[-1] - 1)
 
+    @property
+    def mode(self):
+        return jnp.argmax(self.logits, axis=-1).astype(jnp.int32)
+
+    def entropy(self):
+        log_probs = jax.nn.log_softmax(self.logits)
+        return -jnp.sum(mul_exp(log_probs, log_probs), axis=-1)
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py:483]: why not using `log_probs * self.probs` for simplicity?

File: numpyro/distributions/discrete.py [hunk 2]
```diff
@@ -143,6 +169,31 @@ def mean(self):
...
+            entropy: The entropy of the Bernoulli distribution.
+        """
+        q = _to_probs_bernoulli(-1.0 * self.logits)
+        p = self.probs
+        logq = -jax.nn.softplus(self.logits)
+        logp = -jax.nn.softplus(-1.0 * self.logits)
+
+        # Could make this into a function if we need it elsewhere.
+        qlogq = jnp.where(q == 0.0, 0.0, q * logq)
+        plogp = jnp.where(p == 0.0, 0.0, p * logp)
```
[the remark this answers is not in the document]
Comment (stergiosba) [reply numpyro/distributions/discrete.py:194]: Sure we can.

File: numpyro/distributions/distribution.py [hunk 3]
```diff
@@ -339,14 +343,27 @@ def mean(self):
         """
         Mean of the distribution.
         """
-        raise NotImplementedError
+        raise NotImplementedError(f"mean is not implemented for {self.name}")
```
[the remark this answers is not in the document]
Comment (stergiosba) [reply numpyro/distributions/distribution.py:346]: True

File: numpyro/distributions/discrete.py [hunk 4]
```diff
@@ -102,6 +103,31 @@ def mean(self):
...
+        """Calculates the entropy of the Bernoulli distribution with probability p.
+        H(p,q)=-qlog(q)-plog(p) where q=1-p.
+        With extra care for p=0 and p=1.
+
+        Returns:
+            entropy: The entropy of the Bernoulli distribution.
+        """
+        probs0 = 1.0 - self.probs
+        probs1 = self.probs
+        log_probs0 = jnp.log1p(-1.0 * probs1)
```
[the remark this answers is not in the document]
Comment (stergiosba) [reply numpyro/distributions/discrete.py:123]: I think that was enforced from Black. Maybe I am wrong.

File: numpyro/distributions/discrete.py [hunk 5]
```diff
@@ -102,6 +103,31 @@ def mean(self):
...
+        log_probs1 = jnp.log(probs1)
+
+        # Could make this into a function if we need it elsewhere.
+        qlogq = jnp.where(probs0 == 0.0, 0.0, probs0 * log_probs0)
+        plogp = jnp.where(probs1 == 0.0, 0.0, probs1 * log_probs1)
```
[the remark this answers is not in the document]
Comment (stergiosba) [reply numpyro/distributions/discrete.py:128]: Ok will test.

**Comment by fehiepsi** [comment]
Hi @stergiosba, it seems that using `-logq / p - logits` is pretty stable
```py
import jax
import jax.numpy as jnp

logits = jnp.array([-100., -80., -60., -40., -20., 0., 20., 40., 60.])
logq = -jax.nn.softplus(logits)
p = jax.scipy.special.expit(logits)
p_ = jnp.clip(p, a_min=jnp.finfo(p).tiny)
logq / p_ + logits
```
gives
```
Array([-100.       ,  -81.       ,  -61.       ,  -41.       ,
        -21.       ,   -1.3862944,    0.       ,    0.       ,
          0.       ], dtype=float32)
```
while the actual values are
```py
from decimal import Decimal
for i in logits:
    l = Decimal(str(i))
    p = 1 / (1 + (-l).exp())
    print("entropy", -((1 - p) * (1 - p).ln() + p * p.ln()) / p)
```
```
entropy 1E+2
entropy 80.00000000000000000000000002
entropy 61.00496650303780216962340234
entropy 41.00000000000197983295132689
entropy 21.00000000103057681049624664
entropy 1.386294361119890618834464243
entropy 4.328422607333389151485388142E-8
entropy 1.741825244552915897262840401E-16
entropy 5.487531564015271267712575141E-25
```

**Comment by fehiepsi** [comment]
Hi @stergiosba, we will release numpyro 0.14 in a few days. Do you want to incorporate this feature into it?
